src/Vault.sol

```diff
@@ -0,0 +1,175 @@
+
+    ///@dev Emitted on repay()
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling repay()
+    ///@param debtAmount The amount of debt asset being repaid
+    event Repay(address indexed debtAsset, address indexed user, uint debtAmount);
+
+    ///@dev Emitted on withdraw()
+    ///@param collateralAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param collateralAmount The amount of collateral asset withdrawn
+    event Withdraw(address indexed collateralAsset, address indexed user, uint collateralAmount);  
+
+    ///@dev Emitted on liquidation()
+    ///@param collateralAsset The address of the debt asset
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param debtToCover The amount of debt the liquidator wants to cover
+    ///@param liquidatedCollateralAmount The amount of collateral received by the liquidator
+    event Liquidation(address indexed collateralAsset, address indexed debtAsset, address indexed user, uint debtToCover, uint liquidatedCollateralAmount);
+    
+    ///@notice To allow for margined collateralization; similar to FX brokers
+    ///@dev Fixed-point number; to set collateralization level of debt engine
+    /*Note: collateralLevel is a fixed-point number representation of a decimal (e.g. 0.8), | collateralLevel = range(0, 1e18]
+    *       therefore value supplied has to be order of magnitude smaller than the precision of collateral asset itself.
+    *       Example: 1 DAI (1e18) | 80%: 0.8 DAI (8e17) -> therefore, collateralLevel = 8e17
+    */     
+    uint public collateralLevel; 
+
+    ///@dev Returns decimal places of price as dictated by Chainlink Oracle
+    uint public immutable scalarFactor;
+
+    
+    constructor(address dai_, address weth_, address priceFeedAddress, uint collateralLevel_) {
+        collateral = IERC20Decimals(weth_);
+        debt = IERC20Decimals(dai_);
+
+        priceFeed = AggregatorV3Interface(priceFeedAddress);
+        scalarFactor = 10**priceFeed.decimals();
+
+        // collateralization level
+        collateralLevel = collateralLevel_;
+    }
+
+    ///@dev Users deposit collateral asset into Vault
+    ///@param collateralAmount Amount of collateral to deposit
+    function deposit(uint collateralAmount) external {       
+        deposits[msg.sender] += collateralAmount;
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");  
+        emit Deposit(address(collateral), msg.sender, collateralAmount);
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for colleteralization calculation
+    ///@param debtAmount Amount of debt asset to borrow
+    function borrow(uint debtAmount) external {             
```

Comment:
The user can borrow several times using the same collateral, the collateralisation check should happen on `debts[msg.sender] + debtAmount`

Reply:
so that's why i had `getMaxDebt()` to check the level of available collateral given deposits and existing debt.
it would serve as a check in borrow

Reply:
You're right, good catch. After removing `maxDebt` I did not transform `borrow()` accordingly to work without it.

Reply:
Maybe I'm turning into a better auditor :D